File: Algorithm_307_Range Sum Query - Mutable.cpp
#include <iostream>
#include <vector>
using namespace std;
class NumArray
{
private:
    struct segmentTreeNode
    {
        int start, end;
        segmentTreeNode *left, *right;
        int sum;
        segmentTreeNode(int s, int e) : start(s), end(e) {}
    };
    segmentTreeNode *root;
    segmentTreeNode *buildTree(int start, int end, vector<int> nums)
    {
        segmentTreeNode *root = new segmentTreeNode(start, end);
        if (start == end) root->sum = nums[start];
        else
        {
            int mid = (start + end) / 2;
            root->left = buildTree(start, mid, nums);
            root->right = buildTree(mid + 1, end, nums);
            root->sum = root->left->sum + root->right->sum;
        }
         return root;
    }
    void update(segmentTreeNode *root, int i, int val)
     {
        if (root->start == root->end) root->sum = val;
        else
        {
            int mid = (root->start + root->end) / 2;
            if (mid >= i) update(root->left, i, val);
            else update(root->right, i, val);
            root->sum = root->left->sum + root->right->sum;
       }
    }
    int sumRange(segmentTreeNode *root, int i, int j)
    {
        if (root->start == i && root->end == j) return root->sum;
        int mid = (root->start + root->end) / 2;
        if (i > mid) return sumRange(root->right, i, j);
        if (j <= mid) return sumRange(root->left, i, j);
        return sumRange(root->left, i, mid) + sumRange(root->right, mid + 1, j);
    }
public:
    NumArray(vector<int> nums)
    {
        if (nums.size()) root = buildTree(0, nums.size() - 1, nums);
        else root = NULL;
    }

    void update(int i, int val)
    {
        if (root) update(root, i, val);
    }

    int sumRange(int i, int j)
    {
        return root ? sumRange(root, i, j) : 0;
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * obj.update(i,val);
 * int param_2 = obj.sumRange(i,j);
 */

File: Algorithm_307_Range Sum Query - Mutable.java
// 方法1：线段树
// https://leetcode-cn.com/problems/range-sum-query-mutable/solution/qu-yu-he-jian-suo-shu-zu-ke-xiu-gai-by-leetcode/

class NumArray {

    private int[] tree;
    private int n;

    public NumArray(int[] nums) {
        if (nums.length > 0) {
            n = nums.length;
            tree = new int[n * 2];
            buildTree(nums);
        }
    }

    private void buildTree(int[] nums) {
        for (int i = n, j = 0;  i < 2 * n; i++, j++) {
            tree[i] = nums[j];
        }
        for (int i = n - 1; i > 0; i--) {
            tree[i] = tree[i * 2] + tree[i * 2 + 1];
        }    
    }
    
    public void update(int index, int val) {
        index += n;
        tree[index] = val;
        while (index > 0) {
            int left = index;
            int right = index;
            if (index % 2 == 0) {
                right = index + 1;
            } else {
                left = index - 1;
            }
            // parent is updated after child is updated
            tree[index / 2] = tree[left] + tree[right];
            index /= 2;
        }
    }
    
    public int sumRange(int left, int right) {
        // get leaf with value 'l'
        left += n;
        // get leaf with value 'r'
        right += n;
        int sum = 0;
        while (left <= right) {
            if ((left % 2) == 1) { // left是树的右子节点，在上一层时，left已经不包含在求和范围内，所以sum += tree[left], left++
                sum += tree[left];
                left++;
            }
            if ((right % 2) == 0) { // right是树的左子节点，在上一层时，right已经不包含在求和范围内，所以sum += tree[right], right--
                sum += tree[right];
                right--;
            }
            // 对于left是树的左子节点和right是树的右子节点的情况，因为left和right各自所在的树的另一个子节点也包含在求和范围内
            // 所以并不在这一层中把结果加到sum上去，而是先left /= 2, right /= 2，来到上一层。在这一层再进行判断
            left /= 2;
            right /= 2;
        }
        return sum;
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * obj.update(index,val);
 * int param_2 = obj.sumRange(left,right);
 */

 

// 方法2：树状数组
// https://leetcode-cn.com/problems/range-sum-query-mutable/solution/guan-yu-ge-lei-qu-jian-he-wen-ti-ru-he-x-41hv/

class NumArray {
    int[] tree;
    int lowbit(int x) {
        return x & -x;
    }
    int query(int x) {
        int ans = 0;
        for (int i = x; i > 0; i -= lowbit(i)) ans += tree[i];
        return ans;
    }
    void add(int x, int u) {
        for (int i = x; i <= n; i += lowbit(i)) tree[i] += u;
    }

    int[] nums;
    int n;
    public NumArray(int[] _nums) {
        nums = _nums;
        n = nums.length;
        tree = new int[n + 1];
        for (int i = 0; i < n; i++) add(i + 1, nums[i]);
    }
    
    public void update(int i, int val) {
        add(i + 1, val - nums[i]);
        nums[i] = val;
    }
    
    public int sumRange(int l, int r) {
        return query(r + 1) - query(l);
    }
}